src/parser/error.rs

```diff
@@ -1,20 +1,48 @@
-use crate::lexer::{TkType, Token};
+use crate::lexer::{Location, TkType, Token};
 use thiserror::Error;
 
 pub type Result<T> = std::result::Result<T, ParseError>;
 
 #[derive(Debug, Error)]
-pub enum ParseError {
-    #[error("{} expected one of {} but got {}", .1.location(), ShowTkTypeList(.0.to_vec()), .1.tk_type())]
+#[error("{location} {err}")]
+pub struct ParseError {
+    location: Location,
+    err: ParseErrorVariant,
+}
+
+#[derive(Debug, Error)]
+pub enum ParseErrorVariant {
+    #[error("expected one of {} but got {}", ShowTkTypeList(.0.to_vec()), .1.tk_type())]
```

Comment:
This is a kind of pattern to share required fields, you would see it in the project everywhere